zio-http/src/main/scala/zhttp/http/middleware/HttpMiddlewares.scala

```diff
@@ -0,0 +1,196 @@
+
+  /**
+   * Sets cookie in response headers
+   */
+  def addCookie(cookie: Cookie): HttpMiddleware[Any, Nothing] =
+    addHeader(Headers.setCookie(cookie))
+
+  /**
+   * Adds the provided header and value to the response
+   */
+  def addHeader(name: String, value: String): HttpMiddleware[Any, Nothing] =
+    patch((_, _) => Patch.addHeader(name, value))
+
+  /**
+   * Adds the provided header to the response
+   */
+  def addHeader(header: Headers): HttpMiddleware[Any, Nothing] =
+    patch((_, _) => Patch.addHeader(header))
+
+  /**
+   * Adds the provided list of headers to the response
+   */
+  def addHeaders(headers: Headers): HttpMiddleware[Any, Nothing] =
+    patch((_, _) => Patch.addHeader(headers))
+
+  def addCookieM[R, E](cookie: ZIO[R, E, Cookie]): HttpMiddleware[R, E] =
+    patchZIO((_, _) => cookie.mapBoth(Option(_), c => Patch.addHeader(Headers.setCookie(c))))
+
+  /**
+   * Add log status, method, url and time taken from req to res
+   */
+  def debug: HttpMiddleware[Console with Clock, IOException] =
+    makeResponseZIO((method, url, _) => zio.clock.nanoTime.map(start => (method, url, start))) {
+      case (status, _, (method, url, start)) =>
+        for {
+          end <- clock.nanoTime
+          _   <- console
+            .putStrLn(s"${status.asJava.code()} ${method} ${url.asString} ${(end - start) / 1000000}ms")
+            .mapError(Option(_))
+        } yield Patch.empty
+    }
+
+  /**
+   * Creates a new middleware using transformation functions
+   */
+  def makeResponse[S](req: (Method, URL, Headers) => S): PartialResponseMake[S] = PartialResponseMake(req)
+
+  /**
+   * Creates a new middleware using effectful transformation functions
+   */
+  def makeResponseZIO[R, E, S](req: (Method, URL, Headers) => ZIO[R, Option[E], S]): PartialResponseMakeZIO[R, E, S] =
+    PartialResponseMakeZIO(req)
+
+  /**
+   * Creates a middleware that produces a Patch for the Response
+   */
+  def patch[R, E](f: (Status, Headers) => Patch): HttpMiddleware[R, E] =
+    makeResponse((_, _, _) => ())((status, headers, _) => f(status, headers))
+
+  /**
+   * Creates a middleware that produces a Patch for the Response effectfully.
+   */
+  def patchZIO[R, E](f: (Status, Headers) => ZIO[R, Option[E], Patch]): HttpMiddleware[R, E] =
+    makeResponseZIO((_, _, _) => ZIO.unit)((status, headers, _) => f(status, headers))
+
+  /**
+   * Removes the header by name
+   */
+  def removeHeader(name: String): HttpMiddleware[Any, Nothing] =
+    patch((_, _) => Patch.removeHeaders(List(name)))
+
+  /**
+   * Runs the effect after the response is produced
+   */
+  def runAfter[R, E](effect: ZIO[R, E, Any]): HttpMiddleware[R, E] =
+    patchZIO((_, _) => effect.mapBoth(Option(_), _ => Patch.empty))
+
+  /**
+   * Runs the effect before the request is passed on to the HttpApp on which the middleware is applied.
+   */
+  def runBefore[R, E](effect: ZIO[R, E, Any]): HttpMiddleware[R, E] =
```

Comment:
Can be moved to `Middleware.scala`. We will need to add a `contramapZIO` operator although. You can implement it using flatMap I think.